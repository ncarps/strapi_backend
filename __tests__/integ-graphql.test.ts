import { createTestClient } from 'apollo-server-testing'
import gql from 'graphql-tag'
import { constructTestServer } from './__utils'
import { createProduct } from '../src/helper'

const { server }: any = constructTestServer({
  context: {
    createProduct: jest.fn(() => async (product) => {
      return {
        data: {
          createProduct: {
            product: {
              id: '1',
              ...product.data,
            },
          },
        },
      }
    }),
    deleteProduct: jest.fn(() => async (product) => {
      return {
        data: {
          deleteProduct: {
            product: {
              id: '1',
              amount: '100',
              description: 'Delete Product',
              name: 'Jonas',
              productCode: '420',
              quantity: '1003',
            },
          },
        },
      }
    }),
    updatedProduct: jest.fn(() => async (product) => {
      return {
        data: {
          createProduct: {
            product: {
              id: '1',
              ...product.data,
            },
          },
        },
      }
    }),
    createOrder: jest.fn(() => async (order) => {
      return {
        data: {
          createOrder: {
            order: {
              id: '1',
              ...order.data,
              products: [
                {
                  id: '1',
                  name: 'Mirana',
                  description: 'Jumong',
                  amount: '100',
                  productCode: 'QWE123',
                  quantity: '1000',
                },
                {
                  id: '2',
                  name: 'Rubick',
                  description: 'Telekinesis',
                  amount: '1000',
                  productCode: '42069',
                  quantity: '10001',
                },
              ],
            },
          },
        },
      }
    }),

    updateOrder: jest.fn(() => async (order) => {
      return {
        data: {
          updateOrder: {
            order: {
              id: '1',
              ...order.data,
              products: [
                {
                  id: '1',
                  name: 'Mirana',
                  description: 'Jumong',
                  amount: '100',
                  productCode: 'QWE123',
                  quantity: '1000',
                },
              ],
            },
          },
        },
      }
    }),
  },
})

describe('Queries', () => {
  //Mutations
  it('create a product', async () => {
    const CREATE_PRODUCT = gql`
      mutation a($input: createProductInput) {
        createProduct(input: $input) {
          product {
            id
            created_at
            updated_at
            name
            description
            amount
            productCode
            quantity
          }
        }
      }
    `

    const { mutate } = createTestClient(server)
    const res = await mutate({
      mutation: CREATE_PRODUCT,
      variables: {
        input: {
          data: {
            name: 'HOTDOG',
            description: 'Hotdog numbawan',
            amount: '123',
            productCode: '420',
            quantity: '100',
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  it('update a product', async () => {
    const UPDATE_PRODUCT = gql`
      mutation a($input: createProductInput) {
        createProduct(input: $input) {
          product {
            id
            created_at
            updated_at
            name
            description
            amount
            productCode
            quantity
          }
        }
      }
    `

    const { mutate } = createTestClient(server)
    const res = await mutate({
      mutation: UPDATE_PRODUCT,
      variables: {
        input: {
          data: {
            name: 'HOTDOG',
            description: 'Hotdog numbawan',
            amount: '123',
            productCode: '420',
            quantity: '100',
          },
        },
      },
    })

    expect(res.errors).toBeUndefined()
    expect(res.data).toMatchObject({
      createProduct: {
        product: {
          id: '1',
          name: 'HOTDOG',
          description: 'Hotdog numbawan',
          amount: '123',
          productCode: '420',
          quantity: '100',
        },
      },
    })
    expect(res).toMatchSnapshot()
  })

  it('delete a product', async () => {
    const DELETE_PRODUCT = gql`
      mutation a($input: deleteProductInput) {
        deleteProduct(input: $input) {
          product {
            id
            name
            description
            amount
            productCode
            quantity
          }
        }
      }
    `

    const { mutate } = createTestClient(server)
    const res = await mutate({
      mutation: DELETE_PRODUCT,
      variables: {
        input: {
          where: {
            id: '1',
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  it('create a order', async () => {
    const CREATE_ORDER = gql`
      mutation a($input: createOrderInput) {
        createOrder(input: $input) {
          order {
            id
            created_at
            updated_at
            orderNumber
            receipientName
            products {
              id
              created_at
              updated_at
              name
              description
              amount
              productCode
              quantity
            }
          }
        }
      }
    `

    const { mutate } = createTestClient(server)
    const res = await mutate({
      mutation: CREATE_ORDER,
      variables: {
        input: {
          data: {
            orderNumber: '42069',
            receipientName: 'Gavin Salcedo',
            products: ['1', '2'],
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  it('Update a order', async () => {
    const UPDATE_ORDER = gql`
      mutation updateOrder($input: updateOrderInput) {
        updateOrder(input: $input) {
          order {
            id
            created_at
            updated_at
            orderNumber
            receipientName
            products {
              id
              created_at
              updated_at
              name
              description
              amount
              productCode
              quantity
            }
          }
        }
      }
    `

    const { mutate } = createTestClient(server)
    const res = await mutate({
      mutation: UPDATE_ORDER,
      variables: {
        input: {
          where: {
            id: '1',
          },
          data: {
            orderNumber: '420699',
            receipientName: 'Nastzzz',
            products: ['1'],
          },
        },
      },
    })
    expect(res).toMatchSnapshot()
  })
})
