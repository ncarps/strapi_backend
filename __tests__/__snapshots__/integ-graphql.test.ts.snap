// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries Update a order 1`] = `
Object {
  "data": Object {
    "updateOrder": Object {
      "order": Object {
        "created_at": null,
        "id": "1",
        "orderNumber": "420699",
        "products": Array [
          Object {
            "amount": "100",
            "created_at": null,
            "description": "Jumong",
            "id": "1",
            "name": "Mirana",
            "productCode": "QWE123",
            "quantity": "1000",
            "updated_at": null,
          },
        ],
        "receipientName": "Nastzzz",
        "updated_at": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries create a order 1`] = `
Object {
  "data": Object {
    "createOrder": Object {
      "order": Object {
        "created_at": null,
        "id": "1",
        "orderNumber": "42069",
        "products": Array [
          Object {
            "amount": "100",
            "created_at": null,
            "description": "Jumong",
            "id": "1",
            "name": "Mirana",
            "productCode": "QWE123",
            "quantity": "1000",
            "updated_at": null,
          },
          Object {
            "amount": "1000",
            "created_at": null,
            "description": "Telekinesis",
            "id": "2",
            "name": "Rubick",
            "productCode": "42069",
            "quantity": "10001",
            "updated_at": null,
          },
        ],
        "receipientName": "Gavin Salcedo",
        "updated_at": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries create a product 1`] = `
Object {
  "data": Object {
    "createProduct": Object {
      "product": Object {
        "amount": "123",
        "created_at": null,
        "description": "Hotdog numbawan",
        "id": "1",
        "name": "HOTDOG",
        "productCode": "420",
        "quantity": "100",
        "updated_at": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries delete a product 1`] = `
Object {
  "data": Object {
    "deleteProduct": Object {
      "product": Object {
        "amount": "100",
        "description": "Delete Product",
        "id": "1",
        "name": "Jonas",
        "productCode": "420",
        "quantity": "1003",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries update a product 1`] = `
Object {
  "data": Object {
    "createProduct": Object {
      "product": Object {
        "amount": "123",
        "created_at": null,
        "description": "Hotdog numbawan",
        "id": "1",
        "name": "HOTDOG",
        "productCode": "420",
        "quantity": "100",
        "updated_at": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
